orchagent/lagorch.cpp

```diff
@@ -0,0 +1,217 @@
+                    continue;
+                }
+
+                assert(port.m_type == Port::PHY_PORT);
+
+                if (addLagMember(lag, port))
+                    it = consumer.m_toSync.erase(it);
+                else
+                    it++;
+            }
+            else if (op == DEL_COMMAND)
+            {
+                assert(m_lags[lag_alias].find(port_alias) != m_lags[lag_alias].end());
+                assert(port.m_type == Port::LAG_MEMBER);
+
+                if (removeLagMember(lag, port))
+                    it = consumer.m_toSync.erase(it);
+                else
+                    it++;
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Unknown operation type %s\n", op.c_str());
+                it = consumer.m_toSync.erase(it);
+            }
+        }
+    }
+}
+
+bool LagOrch::addLag(string lag)
+{
+    sai_object_id_t lag_id;
+    sai_status_t status = sai_lag_api->create_lag(&lag_id, 0, NULL);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to create LAG %s lid:%llx", lag.c_str(), lag_id);
+        return false;
+    }
+
+    SWSS_LOG_ERROR("Create an empty LAG %s lid:%llx", lag.c_str(), lag_id);
+
+    Port p(lag, Port::LAG);
+    p.m_lag_id = lag_id;
+    m_portsOrch->setPort(lag, p);
+
+    m_lags[lag] = set<string>();
+
+    return true;
+}
+
+bool LagOrch::removeLag(Port lag)
+{
+    /* Retry when the LAG still has members */
+    if (m_lags[lag.m_alias].size() > 0)
+    {
+        SWSS_LOG_ERROR("Failed to remove non-empty LAG %s", lag.m_alias.c_str());
+        return false;
+    }
+
+    sai_status_t status = sai_lag_api->remove_lag(lag.m_lag_id);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to remove LAG %s lid:%llx\n", lag.m_alias.c_str(), lag.m_lag_id);
+        return false;
+    }
+
+    SWSS_LOG_ERROR("Remove LAG %s lid:%llx\n", lag.m_alias.c_str(), lag.m_lag_id);
+
+    m_lags.erase(lag.m_alias);
+    m_portsOrch->removePort(lag.m_alias);
+
+    return true;
+}
+
+bool LagOrch::addLagMember(Port lag, Port port)
+{
+    sai_attribute_t attr;
+    vector<sai_attribute_t> attrs;
+
+    attr.id = SAI_LAG_MEMBER_ATTR_LAG_ID;
+    attr.value.oid = lag.m_lag_id;
+    attrs.push_back(attr);
+
+    attr.id = SAI_LAG_MEMBER_ATTR_PORT_ID;
+    attr.value.oid = port.m_port_id;
+    attrs.push_back(attr);
+
+    sai_object_id_t lag_member_id;
+    sai_status_t status = sai_lag_api->create_lag_member(&lag_member_id, attrs.size(), attrs.data());
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to add member %s to LAG %s lid:%llx pid:%llx\n",
+                port.m_alias.c_str(), lag.m_alias.c_str(), lag.m_lag_id, port.m_port_id);
+        return false;
+    }
+
+    SWSS_LOG_ERROR("Add member %s to LAG %s lid:%llx pid:%llx\n",
+            port.m_alias.c_str(), lag.m_alias.c_str(), lag.m_lag_id, port.m_port_id);
+
+    port.m_type = Port::LAG_MEMBER;
```

Comment:
the Port::LAG_MEMBER and Port::PHY_PORT are different attribute for a physical port. should not be exclusive to each other. A physical port can be both PHY_PORT and LAG_MEMBER.